File: backend/app/db/models/channel.py
"""Channel model for Telegram channels."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Index, Integer, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.db.models.enums import BackfillMode, DesignerSource, DownloadMode, TitleSource

if TYPE_CHECKING:
    from app.db.models.design_source import DesignSource
    from app.db.models.import_source import ImportSource
    from app.db.models.job import Job
    from app.db.models.telegram_message import TelegramMessage


class Channel(Base):
    """Represents a monitored Telegram channel or virtual import source channel (#237)."""

    __tablename__ = "channels"

    # Primary key
    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )

    # Virtual channel flag (#237) - True for import source channels, False for Telegram channels
    is_virtual: Mapped[bool] = mapped_column(Boolean, default=False)

    # Import source link (#237) - Only set for virtual channels
    import_source_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("import_sources.id", ondelete="CASCADE"), nullable=True
    )

    # Telegram identification (nullable for virtual channels)
    telegram_peer_id: Mapped[str | None] = mapped_column(String(64), unique=True, nullable=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    username: Mapped[str | None] = mapped_column(String(255), nullable=True)
    invite_link: Mapped[str | None] = mapped_column(String(512), nullable=True)
    is_private: Mapped[bool] = mapped_column(Boolean, default=False)
    is_enabled: Mapped[bool] = mapped_column(Boolean, default=True)

    # Channel settings
    backfill_mode: Mapped[BackfillMode] = mapped_column(
        Enum(BackfillMode), default=BackfillMode.LAST_N_MESSAGES
    )
    backfill_value: Mapped[int] = mapped_column(Integer, default=100)
    download_mode: Mapped[DownloadMode] = mapped_column(
        Enum(DownloadMode), default=DownloadMode.MANUAL
    )
    download_mode_enabled_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, doc="When download mode was changed from MANUAL"
    )
    library_template_override: Mapped[str | None] = mapped_column(String(512), nullable=True)
    title_source_override: Mapped[TitleSource | None] = mapped_column(
        Enum(TitleSource), nullable=True
    )
    designer_source_override: Mapped[DesignerSource | None] = mapped_column(
        Enum(DesignerSource), nullable=True
    )

    # Sync state
    last_ingested_message_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    last_backfill_checkpoint: Mapped[int | None] = mapped_column(Integer, nullable=True)
    last_sync_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc)
    )

    # Relationships
    messages: Mapped[list[TelegramMessage]] = relationship(
        "TelegramMessage", back_populates="channel", cascade="all, delete-orphan"
    )
    design_sources: Mapped[list[DesignSource]] = relationship(
        "DesignSource", back_populates="channel", cascade="all, delete-orphan"
    )
    jobs: Mapped[list[Job]] = relationship(
        "Job", back_populates="channel", cascade="all, delete-orphan"
    )
    import_source: Mapped[ImportSource | None] = relationship(
        "ImportSource", back_populates="channel"
    )

    # Indexes and constraints
    __table_args__ = (
        Index("ix_channels_enabled_sync", "is_enabled", "last_sync_at"),
        Index("ix_channels_import_source", "import_source_id"),
        Index("ix_channels_virtual", "is_virtual"),
        # Each import source can only have one virtual channel
        UniqueConstraint("import_source_id", name="uq_channels_import_source"),
    )

File: frontend/src/types/channel.ts
// Must match backend/app/db/models/enums.py exactly
export type BackfillMode = 'ALL_HISTORY' | 'LAST_N_MESSAGES' | 'LAST_N_DAYS'
export type DownloadMode = 'DOWNLOAD_ALL' | 'DOWNLOAD_ALL_NEW' | 'MANUAL'
export type TitleSource = 'CAPTION' | 'FILENAME' | 'MANUAL'
export type DesignerSource = 'CAPTION' | 'CHANNEL' | 'MANUAL'

export interface Channel {
  id: string
  telegram_peer_id: string
  title: string
  username: string | null
  invite_link: string | null
  is_private: boolean
  is_enabled: boolean
  backfill_mode: BackfillMode
  backfill_value: number
  download_mode: DownloadMode
  library_template_override: string | null
  title_source_override: TitleSource | null
  designer_source_override: DesignerSource | null
  last_ingested_message_id: number | null
  last_backfill_checkpoint: number | null
  last_sync_at: string | null
  created_at: string
  updated_at: string
}

export interface ChannelCreate {
  title: string
  username?: string
  invite_link?: string
  is_private?: boolean
  telegram_peer_id?: string
  // Backfill settings (used for update after create)
  backfill_mode?: BackfillMode
  backfill_value?: number
  start_backfill?: boolean // UI-only: trigger backfill after creation
}

export interface ChannelUpdate {
  title?: string
  username?: string
  invite_link?: string
  is_private?: boolean
  is_enabled?: boolean
  backfill_mode?: BackfillMode
  backfill_value?: number
  download_mode?: DownloadMode
  library_template_override?: string | null
  title_source_override?: TitleSource | null
  designer_source_override?: DesignerSource | null
}

export interface ChannelList {
  items: Channel[]
  total: number
  page: number
  page_size: number
  pages: number
}

// Download mode preview/update types (for confirmation flow)
export interface DownloadModePreviewResponse {
  channel_id: string
  current_mode: DownloadMode
  new_mode: DownloadMode
  designs_to_queue: number
}

export interface DownloadModeRequest {
  download_mode: DownloadMode
  confirm_bulk_download?: boolean
}

export interface DownloadModeResponse {
  channel_id: string
  download_mode: DownloadMode
  designs_queued: number
}
